### DISCUSSION TITLE
[동현] State Management 회고록

### DISCUSSION THREAD

**Discussion by D5ng** [React-Core-Learn/State-Management#4] [회고록]
## 느낀 점

아직까지도 100%로 Observer Pattern에 대해 이해하지는 못했지만 어디에서 활용할 수 있을지 정도는 충분히 감을 잡은것 같다. 이번 상태 관리 시스템 만들기를 하면서 그동안 동작하는 방식이 궁금해왔던 `Redux`, `Zustand`, `Tanstack Query`, `SWR`등 소스코드를 분석하면서 Observer Pattern과 어느정도 연관이 있다는것을 알았다. Zustand와 SWR 같은 경우 Provider를 사용하지 않는게 신기했었지만, 소스 코드를 봐보니 `useSyncExternalStore`라는 React Hook API를 활용해서 만들었던 것이였다.

결국 상태 관리는 Observer Pattern을 통해 상태 변경을 감지하고 알리는 방식이 중요한 핵심이라고 생각한다.

## 배운 점

- 상태 관리가 어떤 식으로 관리되는지 직접 구현하면서 이해했다. (Observer Pattern)
- Redux와 같은 라이브러리는 리액트의 Context API를 통해 변경된 상태를 감지하여 리렌더링을 발생하도록 한다. Provider를 사용하지 않는 라이브러리라면  `useSyncExternalStore`로 변경된 상태를 감지해 리렌더링을 발생하도록 한다.
- 문법적이지만, `Object.defineProperty`가 어떻게 쓰이는지 알 수 있었다.

**Comment by Seoin02** [comment]
진짜 많은 소스 코드를 분석했군요!! 저도 봐볼게요!
